Merge Top100 songs statistics to Develop

Ronchoke opened the pull request.

Ronchoke committed:
add .gitignore file
inserted pycharm metadata files not relevant to project.

Ronchoke committed:
Create new module for parsing the top 100 songs
Add get_full_url_to_lyrics function
Add .pytest_cache into .gitignore

Ronchoke committed:
Get Lyrics for Song Title.
Add get_lyrics function.

Ronchoke committed:
Create word counting function.
Add count_words_in_text function.

Ronchoke committed:
Parsing for top 100 chart feature.
Add get_top_100_songs function.
TODO: featuring artists raise error.

Ronchoke committed:
Fixed get_full_url_to_lyrics
Add find_song_title_link function
Add get_content_from_url

Ronchoke committed:
Creat feature for merging dictionaries.
The merge_all_word_counts takes two dictionary type inputs and merge
into the latter and returns it.
The merge_dictionaries_by_key function calls the merge_all_word_counts.

Ronchoke committed:
Plot the most used words in the top 100 ranked songs.
Add plot_most_used_words.

Ronchoke committed:
Add Statistics Plots for Collected Data
Plot Most Verbal Artists
Plot Most Used Words

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+"""This code finds the top 100 most popular songs by parsing and without using OOP
```
Not code - Module

Ronchoke replied:
Done.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+
+def get_content_from_url(url: str) -> Soup:
+    """ This function requests the content of a given HTML url,
+     parses it and returns a list of all relevant content under the text_specifier.
```
What is a text_specifier ?

Ronchoke replied:
Documentation was updated to elaborate on context

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+def get_content_from_url(url: str) -> Soup:
+    """ This function requests the content of a given HTML url,
+     parses it and returns a list of all relevant content under the text_specifier.
+    :param url: str
```
Do ellaborate

Ronchoke replied:
This wasn't correct, changed the description. also return type on line 18 to be BeautifulSoup, wasn't sure if it should be Soup or BeautifulSoup in the function typing on line 14. What do you think?

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    """
+    # Connect to url and check response
+    req = requests.get(url)
+    if req.status_code != 200:
```
200 is a magic number. Don't use magic number.
What you really want to do is to assign it in a proper location.
If this number is being used in sevral places throught the module make it a global parameter. If you are using it in this function's context only, create a local paramter.
in both cases it must be upper cased, MY_PARAMETER = 200

Ronchoke replied:
changed, though not in upper cased letters as the pycharm IDE claims it's not on standard

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    # Connect to url and check response
+    req = requests.get(url)
+    if req.status_code != 200:
+        print(url)
```
You don't need to use the print funcion.
Instead you can do it using the raise function.
`raise ConnectionError(f'Invalid URL: {url}')`

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+
+def find_song_title_link(song_title: str, songs: list) -> str:
+    """This function iterates through HTML syntax from https://www.lyrics.com/ website
+    parsed by the get_content_from_url function
```
Don't mention how it got parsed. If you will change it in the future it will make your documention deprecated and you won't even know.
(which SOLID principle have you violated?)

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    parsed by the get_content_from_url function
+    and find's the relative link to the song_title lyrics.
+    :param song_title: str
+    :param songs: list of HTML syntax content
```
Not indicative name. if its a list call it perhaps songs_list

Ronchoke replied:
Changed to songs_list.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    """
+    for song in songs:
+        find_song_title = song.a.text.lower()
+        if '(' in find_song_title:
```
the '(' is also a magic. Fix it as before..
Fix everywher , and every magic

Ronchoke replied:
This doesn't make sense here as it is a single character.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+            find_song_title = find_song_title.split(' (')[0]
+        if song_title.lower() == find_song_title:
+            return song.find('a').attrs['href']
+    return ''
```
?

Ronchoke replied:
what is the comment about?

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+        if '(' in find_song_title:
+            find_song_title = find_song_title.split(' (')[0]
+        if song_title.lower() == find_song_title:
+            return song.find('a').attrs['href']
```
It is preferable to rueturn from one place only inside a fucntion. This way it is easier to debug and to stay on top of things.

Ronchoke replied:
If it is just a preference I am leaving it as is.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+def get_full_url_to_lyrics(song_title: str, artist: str) -> str:
+    """Gets the full link to the lyric's page for a song_title and artist name,
+    from https://www.lyrics.com website and return all the lyrics.
+    :param song_title: str.
```
Again. ellaborate on the function's variables.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    :return: url: str
+    """
+    # Check Input
+    if type(song_title) is not str or type(artist) is not str:
```
try using the functions any and all.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+        raise ValueError("Artist\'s Name must be given")
+
+    # Handle multiple Artists
+    if "With" in artist:
```
Magics

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+
+    # Create Artist page URL address
+    base = 'https://www.lyrics.com/'
+    url = base + 'artist/' + artist.lower().replace(' ', '+')
```
using + to connect strings is unneat. use format/f

Ronchoke replied:
I dissagree, it is less readable, but I changed it non-the-less

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    return ''
+
+
+def get_full_url_to_lyrics(song_title: str, artist: str) -> str:
```
The funcion is way to lomg amd complex. Refator it.

Ronchoke replied:
did my best, added few functions to shorten it.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    if type(text) is not str:
+        raise ValueError("Input must be string")
+
+    # Clean rows from Text
```
Strip*

Ronchoke replied:
Changed

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+    Billboard hot-100 page (https://www.billboard.com/charts/hot-100)
+    :return: songs_dict: dict
+    """
+    url = 'https://www.billboard.com/charts/hot-100'
```
URLs are also magics. Upper case it and pick a better name.

Ronchoke replied:
renamed it yet the uppercase thing is just raising problems in pycharm ide

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+
+def merge_dictionaries_by_key(songs_dict: dict, key: any) -> dict:
+    """This function runs through a dictionary with values of type dictionary,
+    and merges the values's values by the key given (value[key]).
```
values's values?

Ronchoke replied:
This function was updated completely, as it was missing the actual code to take care different types of values.
Documentation, and code changed.

Maorsou commented on Top100/top100songs.py:
```diff
@@ -0,0 +1,257 @@
+
+def plot_most_used_words(tot_word_count: dict) -> plt:
+    """
+    This function gets a dictionary of word: counts of the top 100 ranked songs lyrics.
```
dictionary of word:?

Ronchoke replied:
changed documentation to both plot function to describe more accurately it's content.

Ronchoke committed:
Improved Documentation and future TODO comments.
get_content_from_url updated for Magic number 200 as approved_request.

Ronchoke committed:
Refactoring get_full_url_to_lyrics
Removed input validation and manipulation into new functions:
Add check_song_title_input function
Add check_artist_input function

Ronchoke committed:
Refactor get_top_100_songs for readability
Add insert_data_to_songs_dict function

Ronchoke committed:
Update documentations for plotting functions
Add description for plot_most_used_words general description and
params and return information.
Add description for plot_most_verbal_artist general description and
params and return information.

Ronchoke committed:
Modify merge_dictionaries_by_key
Updated documentation.
Added merge for non dictionary type values.

Ronchoke committed:
Refactor get_full_url_to_lyrics
Change of string addition to format.

Ronchoke committed:
Update validate_artist_input for duplications
Added a general statment for separating magic words handling.

Ronchoke committed:
Update to find_song_title_link
Changed variable name songs to songs_list
Corrected documentation accordingly to changes, and generalized the
description.

Ronchoke committed:
Updated get_content_from_url documentation
Correct return type of BeautifulSoup object.

Ronchoke committed:
Update all documentations and comment
Added descriptions to all params and return values, added explenations
and clarifications in general description.

Ronchoke committed:
Update README.md
